node/src/types/deploy.rs

```diff
@@ -332,6 +369,40 @@ impl Deploy {
         &self.session
     }
 
+    /// Returns true iff:
+    ///   * the deploy hash is correct (should be the hash of the header), and
+    ///   * the body hash is correct (should be the hash of the body), and
+    ///   * the approvals are all valid signatures of the deploy hash
+    ///
+    /// Note: this is a relatively expensive operation, requiring re-serialization of the deploy,
+    ///       hashing, and signature checking, so should be called as infrequently as possible.
+    pub fn is_valid(&self) -> bool {
```

Comment:
We should also have `is_valid()` check on the `DeployHeader`, to verify it's correct when we're downloading it as dependencies, before we even download the whole block. I remember you saying it is there, can you point me at it?

Reply:
Well, it was previously part of the deserialize method on `Deploy`, so if the deploy hash was invalid we'd abandon deserializing the rest.  We now don't want validation to happen during deserialization, so there's not really a way to avoid downloading and deserializing the full deploy before running the validation check.

Is there currently anywhere we only ask a peer for a deploy header?  If so, then I agree we should have an `is_valid()` for the header too, but if not, then I don't see a need to add it.

Reply:
It's the responsibility of `Fetcher` to use as little bandwidth as possible. From the consensus point of view, it's interested in specific deploy. It will ask the fetcher to download deploy with hash `abcdef123` and fetcher should first download the header, see if it's hash matches the one consensus asked for and if so then continue downloading the body. 

The same should be happening for the block (there's no body there yet though).

Reply:
But is that actually happening?  And is it really worthwhile?  I would think it isn't normally useful - we'd be optimising for the rare case when a deploy or block is invalid.

Reply:
> But is that actually happening?

IDK, you tell me. I didn't write the `Fetcher` :) 

>  And is it really worthwhile? I would think it isn't normally useful - we'd be optimising for the rare case when a deploy or block is invalid.

I doubt it will be a rare case when the network becomes valuable. You had experienced yourself that spamming one "corner" of the node slows down the whole thing.

Reply:
> IDK, you tell me. I didn't write the Fetcher :)

:D  OK, well having looked now, I don't believe it is doing that.

> I doubt it will be a rare case when the network becomes valuable. You had experienced yourself that spamming one "corner" of the node slows down the whole thing.

Yes, I agree about having experienced it, but I'm still not sure if fetching headers separately from bodies will be how we handle that.  Ultimately, it'd be trivial for an attacker using this approach to just send a valid header followed by an invalid body, in which case we haven't really prevented anything.

I'd still like to leave the `is_valid` method off the header for now until we see if we definitely do need it, especially as there wouldn't be anywhere in the current code where we'd be calling it from.  However, if I'm outvoted I can of course add it.

Reply:
What I was thinking about is something like this:
```
Actor A -> Actor B: NewDeploys(hashA, hashB, hashC, …)
Actor B -> Actor A: GetDeploys(hashB, hashC, hashD, …, hashZXY)
Actor A -> Actor B: Opens stream:
  <start_stream>
     <start><deployB_header><deployB_body><end>
     <start><deployC_header><deployC_body><end>
     …
  <end_stream>
```
and when `Actor B` after downloading (for example)  `<deployB_header>` detect hash mismatch it disconnects from the sender.